AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m

```diff
@@ -46,97 +47,106 @@ + (void)performDbTokenAsyncOperationWithHttpClient:(id<MSHttpClientProtocol>)htt
+              NSError *serializeError = [[NSError alloc]
+                  initWithDomain:kMSACDataStoreErrorDomain
+                            code:MSACDataStoreErrorJSONSerializationFailed
+                        userInfo:@{
+                          NSLocalizedDescriptionKey :
+                              [NSString stringWithFormat:@"Can't deserialize tokens with error: %@", [tokenResponsejsonError description]]
+                        }];
+              completionHandler([[MSTokensResponse alloc] initWithTokens:nil], serializeError);
+              return;
+            }
+            if ([(NSArray *)jsonDictionary[kMSTokens] count] == 0) {
+              MSLogError([MSDataStore logTag], @"Invalid token exchange service response.");
+              NSError *errorResponse = [[NSError alloc]
+                  initWithDomain:kMSACDataStoreErrorDomain
+                            code:MSACDataStoreInvalidTokenExchangeResponse
+                        userInfo:@{NSLocalizedDescriptionKey : [NSString stringWithFormat:@"Invalid token exchange service response."]}];
+              completionHandler([[MSTokensResponse alloc] initWithTokens:nil], errorResponse);
+              return;
+            }
+
+            // Create token result object.
+            MSTokenResult *tokenResult = [[MSTokenResult alloc] initWithDictionary:(NSDictionary *)jsonDictionary[kMSTokens][0]];
+
+            // Create token response object.
+            MSTokensResponse *tokensResponse = [[MSTokensResponse alloc] initWithTokens:@[ tokenResult ]];
+
+            // Token exchange did not get back an error but acquiring the token did not succeed either
+            if (tokenResult && ![tokenResult.status isEqualToString:kMSTokenResultSucceed]) {
+              MSLogError([MSDataStore logTag], @"Token result had a status of %@", tokenResult.status);
+              NSError *statusError = [[NSError alloc]
+                  initWithDomain:kMSACDataStoreErrorDomain
+                            code:MSACDataStoreErrorHTTPError
+                        userInfo:@{
+                          NSLocalizedDescriptionKey : [NSString stringWithFormat:@"Token result had a status of %@", tokenResult.status]
+                        }];
+              completionHandler(tokensResponse, statusError);
+              return;
+            }
+
+            // Cache the newly acquired token.
+            [MSTokenExchange saveToken:tokenResult];
+            completionHandler(tokensResponse, error);
+          }];
+    } else {
+      MSLogError([MSDataStore logTag], @"No cached token result found, and device is offline.");
+      NSError *error =
+          [[NSError alloc] initWithDomain:kMSACDataStoreErrorDomain
+                                     code:MSACDataStoreUnableToGetTokenError
+                                 userInfo:@{NSLocalizedDescriptionKey : @"No cached token result found, and device is offline."}];
```

Comment:
I have a feedback about the error message used in this class. We're referring cosmos Db token as "token result" everywhere in this class, when it should be called "cosmos Db token". 😄

Reply:
this is because it's not am actual cosmos db token :)
an actual cosmos db token is like `YQ==`
but this token result is ```{dbAccount: xxx, partition: xxxx, token: `YQ==`}` it's generated by our token exchange service~ :) If you think we should change it, should be quick!~